integration/internal/container/ops.go

```diff
@@ -237,3 +237,15 @@ func WithRuntime(name string) func(*TestContainerConfig) {
 		c.HostConfig.Runtime = name
 	}
 }
+
+// WithCDIDevices sets the CDI devices to use to start the container
+func WithCDIDevices(cdiDeviceNames ...string) func(*TestContainerConfig) {
+	return func(c *TestContainerConfig) {
+		request := containertypes.DeviceRequest{
+			Driver:       "cdi",
+			Capabilities: [][]string{{"cdi"}},
```

Comment:
We should fix the implementation to not required capabilities, maybe?
Definitely makes the API awkward.

Reply:
Yes, I was thinking the same thing. The issue is that the `Capabilities` sets are used to forward a `DeviceRequest` to the relevant handler:
```
func (daemon *Daemon) handleDevice(req container.DeviceRequest, spec *specs.Spec) error {
	if req.Driver == "" {
		for _, dd := range deviceDrivers {
			if selected := dd.capset.Match(req.Capabilities); selected != nil {
				return dd.updateSpec(spec, &deviceInstance{req: req, selectedCaps: selected})
			}
		}
	} else if dd := deviceDrivers[req.Driver]; dd != nil {
		if selected := dd.capset.Match(req.Capabilities); selected != nil {
			return dd.updateSpec(spec, &deviceInstance{req: req, selectedCaps: selected})
		}
	}
	return incompatibleDeviceRequest{req.Driver, req.Capabilities}
}
```
Would adding a special case for `req.Driver == "cdi"` here be in order (skipping the `Capabilities` check in that case)?

Note that the only other use of this code is the handler for the `--gpu` flag, which we expect to be replaced by this functionality as soon as it is considered stable (see #45192), I don't see too much risk in adding CDI-specific logic here and then improving the implementation in a follow-up.

What I can see being done is having each driver implement a `Match([][]string ) []string` function and delegate this logic there. This belongs in a follow-up though.